wp/lib/bap_wp/src/z3_utils.ml

```diff
@@ -132,27 +132,38 @@ let mk_smtlib2_compare ?(name = None) (env1 : Env.t) (env2 : Env.t)
   let ctx = Env.get_context env1 in
   mk_smtlib2 ~name ctx smtlib_str declsym
 
-let mk_smtlib2_single ?(name = None) (env : Env.t) (smt_post : string) : Constr.t =
+let mk_smtlib2_single_with_vars
+      ?(name = None) (env : Env.t) (smt_post : string)
+    : Constr.t * string list * string list =
   let var_map = Env.get_var_map env in
   let init_var_map = Env.get_init_var_map env in
-  let smt_post =
-    smt_post
-    |> tokenize
-    |> List.map ~f:(fun token ->
-        match get_z3_name var_map token Var.name with
-        | Some n -> n |> Expr.to_string
+  let tokens = tokenize smt_post in
+
+  (* Here we are replacing variables in the user's provided input with the
+     appropriate z3 variable name.  We also save these variables for user
+     by the caller. *)
+  let (tokens,pre_vars,post_vars) =
+    List.fold_right tokens ~init:([],[],[]) ~f:(fun t (ts,pre_v,post_v) ->
+        match get_z3_name var_map t Var.name with
+        | Some n -> (Expr.to_string n :: ts, pre_v, t :: post_v)
         | None ->
           begin
-            match get_z3_name init_var_map token (fun v -> "init_" ^ Var.name v) with
-            | Some n -> n |> Expr.to_string
-            | None -> token
+            match get_z3_name init_var_map t
+                    (fun v -> "init_" ^ Var.name v) with
```

Comment:
We might want to clean the name here too.

Reply:
I guess I don't totally understand what name cleaning does for us.  It looks like the clean function just drops the `|` character.  I'm guessing that's because z3 doesn't like it?  But for the case of a user-provided specification, if they've used this character wouldn't it be better to have z3 fail loudly than to silently assume, for example, they meant `RAX` instead of `RA|X`?

Reply:
Makes sense. We mostly clean those so Boolector doesn't complain. The case I was thinking about was when a user has a virtual variable in their spec, but your last comment convinced me otherwise. If a user specified a virtual variable in their spec, the variable they wrote wouldn't match what is already in the environment since we clean those names there. If there are no virtual variables, then this is no problem, and we should keep it like this.